### DISCUSSION TITLE
Goreleaser does not output erros of failing command

### DISCUSSION THREAD

**Discussion by Kovah** [goreleaser/goreleaser#1696] [Q&A] [answered]
Hi,
I am currently rewriting one of my tools with Go and found Goreleaser trough Hugo. It seems to be the perfect tool to automate binary releases, but I can't make it work correctly.

The source is located at [Kovah/DevLorem](https://github.com/Kovah/DevLorem/tree/go-rewrite).

- The project uses Rice to bundle static files into the binary, which works great on my machine.
- I am using the official Github Action to run Goreleaser based on the information given in the docs.
- Goreleaser is included as a "test" task with `release --snapshot --skip-publish --rm-dist` to make sure a binary would be compiled correctly.
- In this test task I specified `rice embed-go` as a global pre hook, as it's required to be able to build the final binary.
- The `rice embed-go` command fails, but there is no error thrown, so I am really not sure what the problem is.
- Enabling the debug mode does not help.

This is the output:
```
Run goreleaser/goreleaser-action@v2
✅ GoReleaser version found: v0.140.0
⬇️ Downloading https://github.com/goreleaser/goreleaser/releases/download/v0.140.0/goreleaser_Linux_x86_64.tar.gz...
📦 Extracting GoReleaser...
/bin/tar xz --warning=no-unknown-keyword -C /home/runner/work/_temp/5bf588f9-e3e8-409f-86b0-aa1dd80cae03 -f /home/runner/work/_temp/d1860065-60ab-494b-8926-dbe9d66e83d3
⚠️ No tag found for commit '1450918'. Snapshot forced
🏃 Running GoReleaser...
/opt/hostedtoolcache/goreleaser-action/0.140.0/x64/goreleaser release --snapshot --skip-publish --rm-dist

   • releasing...     
   • loading config file       file=.goreleaser.yml
   • running before hooks
      • running go mod download
      • running go get github.com/GeertJohan/go.rice
      • running rice embed-go
   ⨯ release failed after 8.30s error=hook failed: rice embed-go

##[error]The process '/opt/hostedtoolcache/goreleaser-action/0.140.0/x64/goreleaser' failed with exit code 1
```

What am I doing wrong? :(
There is an example in the docs which includes `rice embed-go` as a pre hook, but nothing else is specified on how to make it work.

Thanks in advance for the help!

**Comment by caarlos0** [comment, marked as answer]
it does print, seems like the command returned and empty output...

https://github.com/goreleaser/goreleaser/blob/master/internal/pipe/before/before.go#L42

**Comment by Kovah** [comment]
Thank you very much for the super fast reply!

I thought that it might be the fault of the rice command not returning an output, so I tested a few other things by running the action workflow locally with act. I am not sure if this is intended, but other commands do not return any output too, if they fail.

```
    • releasing...
    • loading config file       file=.goreleaser.yml
    • running before hooks
       • running which go
    ⨯ release failed after 0.01s error=hook failed: which go
```

To make sure there is no issue with my configuration, I re-initialized Goreleaser with `goreleaser init`.
`go mod download` does not run successfully, but again no output _why_ the program fails to execute.

```
 [command]/opt/hostedtoolcache/goreleaser-action/0.140.0/x64/goreleaser release --debug --snapshot --skip-publish --rm-dist

    • debug logs enabled
    • releasing...
    • loading config file       file=.goreleaser.yml
    • loaded config file        config={ [] {   false false   [] []} [] [] { {  } { }      [] } [{ [] [] [] [] [] []   [] []  {[] []} [CGO_ENABLED=0]  [] []  false}] [{ []  map[386:i386 amd64:x86_64 darwin:Darwin linux:Linux windows:Windows]  []  []}] [] [] {{{ .Tag }}-next} {checksums.txt } [] [] [] [] [] {{[^docs: ^test:]} asc false}  [] {  } {[go mod download go generate ./...]} {  false} { [] [] [] [] [] []   [] []  {[] []} []  [] []  false} {   false} {  false} { false}}
    • parallelism: 4
    • running before hooks
       • running go mod download
       •
    ⨯ release failed after 0.00s error=hook failed: go mod download

[Test with Go/preview-release] ::error::The process '/opt/hostedtoolcache/goreleaser-action/0.140.0/x64/goreleaser' failed with exit code 1
```

Is this a possible bug?

**Reply by caarlos0** [reply]
very weird, I just tried here and it works as expected:

```console
$ ls -lAh
total 24
-rw-r--r--  1 carlos  wheel    32B Jul 21 18:54 .goreleaser.yml
-rwxr-xr-x  1 carlos  wheel    34B Jul 21 18:53 foo.sh
-rw-r--r--  1 carlos  staff    66B Jul 21 18:54 main.go

$ cat .goreleaser.yml
before:
  hooks:
    - ./foo.sh

$ cat foo.sh
#!/bin/sh
echo some output
exit 1

$ cat main.go
package main

import "fmt"

func main() {
	fmt.Println("lalal")
}

$ goreleaser --rm-dist --snapshot

   • releasing...
   • loading config file       file=.goreleaser.yml
   • running before hooks
      • running ./foo.sh
   ⨯ release failed after 0.01s error=hook failed: ./foo.sh
some output

```

my local goreleaser version:

```console
$ goreleaser --version

goreleaser version 0.140.0
commit: 8c93af20f8d8326c621aca2bfc61f9309aac78df
built by: homebrew

```

can you try with this example above?

**Reply by Kovah** [reply]
I tried to run Goreleaser locally and this seems to work, but only if _the script itself_ outputs something.

fail.sh exists:
```
   • running before hooks
      • running go mod download
      •
      • running go generate ./...
      •
      • running ./fail.sh
      • some output

   ⨯ release failed after 0.40s error=hook failed: ./fail.sh
some output
```

In case fail.sh does not exist, I would expect output like `no such file or directory: ./fail.sh`, but I get nothing.

```
   • running before hooks
      • running go mod download
      •
      • running go generate ./...
      •
      • running ./fail.sh
      •
   ⨯ release failed after 0.45s error=hook failed: ./fail.sh

```

I think it would be helpful if OS-level fails are captured too, and not only script output.

**Reply by caarlos0** [reply]
oh, yeah, that makes sense... will fix :) 

thanks

**Reply by caarlos0** [reply]
https://github.com/goreleaser/goreleaser/pull/1698

**Reply by Kovah** [reply]
Thank you very much for the fast support with this!
